### ISSUE TITLE
Succeded with intel-i915-dkms bakckport for LTS61 on 11800h laptop, host suspends and wakes perfectly!

### ISSUE THREAD

**Issue by right-0903** [strongtz/i915-sriov-dkms#101]
I have tried with the dkms module from the repositroy(with 6.1.44-MANJARO or 6.4.10-zen2), and I also tried to build this [6.1.38](https://github.com/intel/linux-intel-lts/releases/tag/lts-v6.1.38-linux-cve-230731T034035Z) from linux-intel-lts with this [aur package](https://aur.archlinux.org/packages/linux-intel-lts-sriov). None of them work with my instance.

Then I found the backport driver ocassionally. You will get the introduction about backport [here](https://github.com/intel/media-delivery/blob/master/doc/intel-gpu-dkms.rst), aka the [out-of-tree driver](https://dgpu-docs.intel.com/driver/kernel-driver-types.html#differences-between-the-out-of-tree-driver-and-upstream-kernel). And you will find more functions about sriov in [backport driver](https://github.com/intel-gpu/intel-gpu-i915-backports/blob/backport/main/drivers/gpu/drm/i915/i915_sriov.c) compared  with [linux-intel-lts](https://github.com/intel/linux-intel-lts/blob/6.1/linux/drivers/gpu/drm/i915/i915_sriov.c), that is no wonder why backport driver work for 11800h while linux-intel-lts not. 

you can install the dkms module similarly, make sure you are 515LTS kernel as backport required, I just tested with it. If you are ubuntu, suse, or redhat refer [this](https://dgpu-docs.intel.com/driver/installation.html#ubuntu-install-steps). There is a packaged dkms module.

Today(10/07/2023), I have packaged for backports driver, you can easily install on arch, [refer this](https://aur.archlinux.org/packages/intel-i915-backport-dkms)

The source doesn't contain a dkms configuration file, you should create one, I list below(I built the debian package in my cloud instance, which contains this.), actually you can download it from intel's repository accroding to [this](https://dgpu-docs.intel.com/driver/installation.html#ubuntu-package-repository).

```bash
PACKAGE_NAME="intel-i915-dkms"
PACKAGE_VERSION="1.23.6.24.230425.29"
AUTOINSTALL="yes"

BUILT_MODULE_NAME[0]="compat"
BUILT_MODULE_LOCATION[0]="compat"
DEST_MODULE_LOCATION[0]="/updates"

BUILT_MODULE_NAME[1]="i915"
BUILT_MODULE_LOCATION[1]="drivers/gpu/drm/i915"
DEST_MODULE_LOCATION[1]="/updates"

BUILT_MODULE_NAME[2]="i915_spi"
BUILT_MODULE_LOCATION[2]="drivers/gpu/drm/i915"
DEST_MODULE_LOCATION[2]="/updates"

BUILT_MODULE_NAME[3]="iaf"
BUILT_MODULE_LOCATION[3]="drivers/gpu/drm/i915/fabric"
DEST_MODULE_LOCATION[3]="/updates"

BUILT_MODULE_NAME[4]="mei"
BUILT_MODULE_LOCATION[4]="drivers/misc/mei/"
DEST_MODULE_LOCATION[4]="/updates"

BUILT_MODULE_NAME[5]="mei-me"
BUILT_MODULE_LOCATION[5]="drivers/misc/mei/"
DEST_MODULE_LOCATION[5]="/updates"

BUILT_MODULE_NAME[6]="mei-gsc"
BUILT_MODULE_LOCATION[6]="drivers/misc/mei/"
DEST_MODULE_LOCATION[6]="/updates"

BUILT_MODULE_NAME[7]="mei_wdt"
BUILT_MODULE_LOCATION[7]="drivers/watchdog/"
DEST_MODULE_LOCATION[7]="/updates"

BUILT_MODULE_NAME[8]="mei_hdcp"
BUILT_MODULE_LOCATION[8]="drivers/misc/mei/hdcp/"
DEST_MODULE_LOCATION[8]="/updates"

BUILT_MODULE_NAME[9]="mei_pxp"
BUILT_MODULE_LOCATION[9]="drivers/misc/mei/pxp/"
DEST_MODULE_LOCATION[9]="/updates"

BUILT_MODULE_NAME[10]="mei_iaf"
BUILT_MODULE_LOCATION[10]="drivers/misc/mei/iaf/"
DEST_MODULE_LOCATION[10]="/updates"

BUILT_MODULE_NAME[11]="intel_vsec"
BUILT_MODULE_LOCATION[11]="drivers/platform/x86/intel"
DEST_MODULE_LOCATION[11]="/updates"

BUILT_MODULE_NAME[12]="pmt_class"
BUILT_MODULE_LOCATION[12]="drivers/platform/x86/intel/pmt"
DEST_MODULE_LOCATION[12]="/updates"

BUILT_MODULE_NAME[13]="pmt_telemetry"
BUILT_MODULE_LOCATION[13]="drivers/platform/x86/intel/pmt"
DEST_MODULE_LOCATION[13]="/updates"

BUILT_MODULE_NAME[14]="pmt_crashlog"
BUILT_MODULE_LOCATION[14]="drivers/platform/x86/intel/pmt"
DEST_MODULE_LOCATION[14]="/updates"

# Find out how many CPU cores can be use if we pass appropriate -j option to make.
# DKMS could use all cores on multicore systems to build the kernel module.
num_cpu_cores()
{
  if [ -x /usr/bin/nproc ]; then
np=1
if [ "$np" -le "64" ]; then
echo "$np"
else
echo "64"
fi
  else
    echo "1"
  fi
}

# Here is where we could specify use of the defconfig. Right now it uses the kernel configuration from /usr/src/
MAKE="cp defconfigs/i915 .config; 'make' -j$(num_cpu_cores) KLIB=/lib/modules/$kernelver olddefconfig; 'make' -j$(num_cpu_cores) KLIB=/lib/modules/$kernelver"
CLEAN="'make' clean"
```

In the future, you can generate one according to `modules.order`, which will be genereted when you `make defconfig-i915 && make`.

# Time to build the driver out of tree

```bash
git clone https://github.com/intel-gpu/intel-gpu-i915-backports.git 
cd intel-gpu-i915-backports
git checkout backport/main
mv /lib/modules/$(uname -r)/kernel/drivers/gpu/drm/i915/i915.ko.xz /lib/modules/$(uname -r)/kernel/drivers/gpu/drm/i915/i915.ko.xzbak
# copy dkms.conf to this dir, I have attach above, I won't do it here
sudo dkms install -m intel-i915-dkms -v 1.23.6.24.230425.29 -k $(uname -r)
# don't forget add GRUB_CMDLINE_LINUX="intel_iommu=on iommu=pt i915.modeset=1 i915.enable_guc=3 i915.max_vfs=7 i915.reset=1 vfio-pci.enable_sriov=1 split_lock_detect=off" to your /etc/default/grub, 
# don't forget download the lastest [firmware](https://github.com/intel-gpu/intel-gpu-firmware) for you platform, tgl_guc_70.7.0.bin for me.
sudo update-grub
# you may should also regenerate initramfs, it is unnecessary for arch linux without [early kms](https://wiki.archlinux.org/title/Kernel_mode_setting#Early_KMS_start)
# after reboot
sudo -i
# don't do it with sudo echo balabala
echo 1 >  /sys/devices/pci0000:00/0000:00:02.0/sriov_numvfs
# vfio bind VF
virsh nodedev-detach pci_0000_00_02_1
```

The following is my configuration.

# Test Environment:

## host info:

```
OS: Manjaro Linux x86_64
Kernel: 5.15.125-1-MANJARO 
CPU: 11th Gen Intel i7-11800H (16) @ 4.600GHz
GPU: Intel TigerLake-H GT1 [UHD Graphics]
GRUB_CMDLINE_LINUX="intel_iommu=on iommu=pt i915.modeset=1 i915.enable_guc=3 i915.max_vfs=7 i915.reset=1 vfio-pci.enable_sriov=1 split_lock_detect=off" (I think not all parms. are necessary)
```

## virsh or qemu configuration 

I got it from [archwiki](https://wiki.archlinux.org/title/QEMU/Guest_graphics_acceleration#SR-IOV), maybe unnecessary.

```xml
    <hyperv mode="custom">
      ...
      <vendor_id state="on" value="GenuineIntel"/>
      ...
    </hyperv>
```

make sure you don't hidden virtualization,  credit to @pagkly . I think the guest driver need a flag to know itself is a VF. Hiding kvm is allowed in my case.

:warning: **Warning:** do not use!

```xml
    <feature policy="disable" name="hypervisor"/>
```

## guest info: 

Driver suggestion: 4577 or [the latest](https://www.intel.com/content/www/us/en/download/726609/intel-arc-iris-xe-graphics-whql-windows.html), some guides recommend 3222, it doesn't work for me.

|Guest 1| backport  kenel driver | driver from intel-lts or this repository|
|-----|-----------------|---------------|
| 3222 or 3111 | code43 | 0% utilization|
|4577|works|code43|

* guest 1
```
Edition: Windows 10 LTSC
Version: 21H2
OS build: 19044.3324
Guest driver: 4577
```

* guest 2
```
Edition: Windows 10 Pro
Version: 2009
OS build: 19042.631
Guest driver: 4577
```

* guest 3
```
Edition: Windows 10 Pro N
Version: 22H2
OS build: 19045.2965
Guest driver: 4577
```

All my VMs work now with 4577.

-----

# Result

Working normally here with intel-i915-dkms bakckport for LTS515 on my 11800h laptop.
![Screenshot_20230817-121228](https://github.com/strongtz/i915-sriov-dkms/assets/111179762/7ffe9f47-c201-401c-a5a6-0eb6e000326d)
Here is the `intel_gpu_top -d sriov` output under headless.
![IMG_20230817_121114632_HDR](https://github.com/strongtz/i915-sriov-dkms/assets/111179762/7dcd7b09-eb87-4865-9b2c-56c54ba58b4e)

-----
# Suspend

About suspend, the guest will not sleep I think, after I wake host up, I reconnect my guest,  there is no relogin animation, and I can't access sleep info by `powercfg /sleepstudy`. In the host, 

```
[  672.034378] PM: suspend entry (s2idle)
[  672.052260] Filesystems sync: 0.017 seconds
[  673.198593] Freezing user space processes ... (elapsed 0.002 seconds) done.
[  673.200765] OOM killer disabled.
[  673.200766] Freezing remaining freezable tasks ... (elapsed 0.001 seconds) done.
[  673.202039] printk: Suspending console(s) (use no_console_suspend to debug)
[  673.202289] i915 0000:00:02.0: VF1 paused
[  673.683927] ACPI: EC: interrupt blocked
[  675.220518] ACPI: EC: interrupt unblocked
[  675.240016] ------------[ cut here ]------------
[  675.240019] i915 0000:00:02.0: can't update enabled VF BAR0 [mem 0x4010000000-0x4016ffffff 64bit]
[  675.240035] WARNING: CPU: 10 PID: 4781 at drivers/pci/iov.c:923 pci_iov_update_resource+0x141/0x150
< I skip here>
[  675.241947] ---[ end trace c9b201cfd30033c9 ]---
[  675.772752] i915 0000:00:02.0: [drm] GT0: GuC firmware i915/tgl_guc_70.7.0.bin version 70.7.0
[  675.772756] i915 0000:00:02.0: [drm] GT0: HuC firmware i915/tgl_huc_7.9.3.bin version 7.9.3
[  675.772936] vfio-pci 0000:00:02.1: enabling device (0000 -> 0002)
[  675.776967] i915 0000:00:02.0: [drm] GT0: HuC: authenticated!
[  675.777487] i915 0000:00:02.0: [drm] GT0: GUC: submission enabled
[  675.777489] i915 0000:00:02.0: [drm] GT0: GUC: SLPC enabled
[  675.777996] i915 0000:00:02.0: [drm] GT0: GUC: RC enabled
[  676.434191] OOM killer enabled.
[  676.434194] Restarting tasks ... 
[  676.434637] mei_hdcp 0000:00:16.0-b638ab7e-94e2-4ea2-a552-d1c54b627f04: bound 0000:00:02.0 (ops i915_hdcp_component_ops [i915])
[  676.435436] mei_pxp 0000:00:16.0-fbf6fcf1-96cf-4e2e-a6a6-1bab8cbe36b1: bound 0000:00:02.0 (ops i915_pxp_tee_component_ops [i915])
[  676.436328] done.
[  676.451166] PM: suspend exit
[  678.523222] i915 0000:00:02.0: VF1 has completed migration
```

`[  678.523222] i915 0000:00:02.0: VF1 has completed migration` is the key I think, it sotres the state of the vm, you can learn about it [here](https://github.com/intel-gpu/intel-gpu-i915-backports/blob/backport/main/drivers/gpu/drm/i915/i915_sriov.c#L38), compared with [linux-intel-lts one](https://github.com/intel/linux-intel-lts/blob/6.1/linux/drivers/gpu/drm/i915/i915_sriov.c).

EDIT: 
1. revise grub confg update, I mess them up, `sudo grub-mkconfig` -> `sudo update-grub`, or  use `sudo grub-mkconfig -o /boot/grub/grub.cfg`.
2. append several kernel commandline parm.s: `intel_iommu=on iommu=pt i915.modeset=1`, I notice not all people use them, but necessary.
3. add several guest instances, and remove version requirement(21h2+), add credits.
4. add an aur repository for this backports package.

**Comment by pagkly** [comment]
Sorry I want to try this but dkms command returns error '/var/lib/../source/dkms.conf' missing

**Comment by right-0903** [comment]
> Sorry I want to try this but dkms command returns error '/var/lib/../source/dkms.conf' missing

copy my `dkms.conf`, I have already listed it above. looking forward to feedback.

**Comment by pagkly** [comment]
Hi, my distro is Arch and backport successfully built with 5.19 kernel. I could not get dkms to install on 5.15. However, once I reboot, GDM could not load. It shows black screen with blinking underscore.

**Comment by right-0903** [comment]
> Hi, my distro is Arch and backport successfully built with 5.19 kernel. I could not get dkms to install on 5.15. However, once I reboot, GDM could not load. It shows black screen with blinking underscore.

Is `Ctrl+Alt+F2-7` available, have you looked into journal such as `journalctl -rb -1`? BTW, I use sddm-git which is running on Wayland, X11 or Wayland login session work.

**Comment by pagkly** [comment]
Cannot build i915 backport dkms module on kernel 5.15, due to error but I can install it on `5.19.13-arch1-1`. 
I tried switching to sddm-git but it says `Error Internal "Process crashed"` when I looked into `journalctl -rb -1`

**Comment by right-0903** [comment]
> Cannot build i915 backport dkms module on kernel 5.15, due to error but I can install it on `5.19.13-arch1-1`. I tried switching to sddm-git but still cant load sddm it says `Error Internal "Process crashed"` when I looked into `journalctl -rb -1`

I can't find a prebuilt linux515 or linux519 for arch linux, so I can't test for you, can you show me dkms build error log on 515? Make sure your subversion is not too old, mine is `5.15.125`, you can also access the prebuilt `linux515-5.15.125-1-x86_64.pkg.tar.zst` `linux515-headers-5.15.125-1-x86_64.pkg.tar.zst` from manjaro mirror, such as [here](https://ftp.halifax.rwth-aachen.de/manjaro/stable/core/x86_64/).

**Comment by right-0903** [comment]
do some changes, make it built on 6.1, then I get a underscore too, that means i915 don't work, but my Nvidia works when I attach another monitor to it. log tell me there are some unknown symbol, one of the reasons is svm, I commented svm, because thera are some changes to vma, svm code use it. But I am not familiar with it and I can't find any commit about i915 svm in mainline.

**Comment by pagkly** [comment]
I tried following your setup on Manjaro 5.15.125 but could not get sr-iov to work.
When I tried to set VF, it says `Numerical result out of range`.
Might be because I have Iris Xe GPU. My CPU is i5 1135G7.

**Comment by right-0903** [comment]
> When I tried to set VF, it says `Numerical result out of range`.

have you set kernel command line? what is the kernel buffer? This issue is related to max_vfs setting which is logged in kernel buffer

**Comment by pwouik** [comment]
I get echo write error no such file or directory when I echo anything other than 0
cpu:11th Gen Intel(R) Core(TM) i5-1135G7 
linux 5.15.127-1-lts515
gpu: Intel Corporation TigerLake-LP GT2 [Iris Xe Graphics]

**Comment by right-0903** [comment]
> I get "driver does not support SR-IOV configuration via sysfs" in journalctl, I guess it's not there yet for Iris Xe

intel claims the driver support that, make sure you have make it work instead of original one. sysfs is enabled here.

**Comment by right-0903** [comment]
> I get echo write error no such file or directory when I echo anything other than 0
> cpu:11th Gen Intel(R) Core(TM) i5-1135G7 
> linux 5.15.127-1-lts515
> gpu: Intel Corporation TigerLake-LP GT2 [Iris Xe Graphics]
> I get "driver does not support SR-IOV configuration via sysfs" in journalctl, I guess it's not there yet for Iris Xe

you can `modinfo i915` to determine the working driver, which will display driver path and dependencies, one of dependencies is `compat`. BTW, if you have set early kms or some other things, make sure you have regenerated initramfs.

**Comment by pwouik** [comment]
[modinfo i915.txt](https://github.com/strongtz/i915-sriov-dkms/files/12503626/modinfo.i915.txt)
seems correct

**Comment by pwouik** [comment]
I regenerated initramfs, I get 'Numerical result out of range'
max_vfs is correctly set : [systool -vm i915.txt](https://github.com/strongtz/i915-sriov-dkms/files/12504234/systool.-vm.i915.txt)

**Comment by right-0903** [comment]
> I regenerated initramfs, I get 'Numerical result out of range'
> max_vfs is correctly set : [systool -vm i915.txt](https://github.com/strongtz/i915-sriov-dkms/files/12504234/systool.-vm.i915.txt)
> 

have you install the latest firmware for guc? post kernel buffer about i915 too.

**Comment by right-0903** [comment]
> I regenerated initramfs, I get 'Numerical result out of range' max_vfs is correctly set : [systool -vm i915.txt](https://github.com/strongtz/i915-sriov-dkms/files/12504234/systool.-vm.i915.txt)

just look into this. 

i915 parm: `modeset:Use kernel modesetting [KMS] (0=disable, 1=on, -1=force vga console preference [default]) (int)`

yours is -1, please append your kernel commandline with mine, once you succeed, then try to remove, something you don't want to add.

**Comment by right-0903** [comment]
> I tried following your setup on Manjaro 5.15.125 but could not get sr-iov to work. When I tried to set VF, it says `Numerical result out of range`. Might be because I have Iris Xe GPU. My CPU is i5 1135G7.

can you try to add enable KMS `i915.modeset=1` and try again?

**Comment by right-0903** [comment]
please append several kernel commandline parm.s: `intel_iommu=on iommu=pt i915.modeset=1`, I notice they are not used by default, but necessary.

**Comment by pwouik** [comment]
same message with new options
archlinux linux-lts515(aur)
`[    0.000000] Command line: BOOT_IMAGE=/vmlinuz-linux-lts515 root=UUID=5f555fd7-165b-4d9a-b566-e45c86beec81 rw intel_iommu=on iommu=pt i915.modeset=1 i915.enable_guc=3 i915.max_vfs=7 i915.reset=1 vfio-pci.enable_sriov=1 split_lock_detect=off
[    0.040381] Kernel command line: BOOT_IMAGE=/vmlinuz-linux-lts515 root=UUID=5f555fd7-165b-4d9a-b566-e45c86beec81 rw intel_iommu=on iommu=pt i915.modeset=1 i915.enable_guc=3 i915.max_vfs=7 i915.reset=1 vfio-pci.enable_sriov=1 split_lock_detect=off
[    3.869860] i915 0000:00:02.0: Running in SR-IOV PF mode
[    3.869929] i915 0000:00:02.0: [drm] GT count: 1, enabled: 1
[    3.870270] i915 0000:00:02.0: [drm] VT-d active for gfx access
[    3.874347] i915 0000:00:02.0: vgaarb: deactivate vga console
[    3.874401] i915 0000:00:02.0: [drm] Using Transparent Hugepages
[    3.875165] i915 0000:00:02.0: vgaarb: changed VGA decodes: olddecodes=io+mem,decodes=io+mem:owns=io+mem
[    3.876438] i915 0000:00:02.0: Direct firmware load for i915/tgl_dmc_ver2_12.bin failed with error -2
[    3.876443] i915 0000:00:02.0: [drm] Failed to load DMC firmware i915/tgl_dmc_ver2_12.bin. Disabling runtime power management.
[    3.876445] i915 0000:00:02.0: [drm] DMC firmware homepage: https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/i915
[    3.887179] i915 0000:00:02.0: [drm] *ERROR* GT0: GuC firmware i915/tgl_guc_70.7.0.bin: fetch failed -ENOENT
[    3.887187] i915 0000:00:02.0: [drm] GT0: GuC firmware(s) can be downloaded from https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/i915
[    3.888006] i915 0000:00:02.0: [drm] GT0: GuC firmware i915/tgl_guc_70.7.0.bin version 0.0.0
[    3.888110] i915 0000:00:02.0: [drm] *ERROR* GT0: GuC initialization failed -ENOENT
[    3.888113] i915 0000:00:02.0: [drm] *ERROR* GT0: Enabling uc failed (-5)
[    3.888115] i915 0000:00:02.0: [drm] *ERROR* GT0: Failed to initialize GPU, declaring it wedged!
[    3.895277] i915 0000:00:02.0: [drm:add_taint_for_CI [i915]] CI tainted:0x9 by intel_gt_init+0xd8/0x600 [i915]
[    3.992730] mei_hdcp 0000:00:16.0-b638ab7e-94e2-4ea2-a552-d1c54b627f04: bound 0000:00:02.0 (ops i915_hdcp_component_ops [i915])
[    4.320866] i915 0000:00:02.0: [drm:add_taint_for_CI [i915]] CI tainted:0x9 by i915_gem_init+0x148/0x240 [i915]
[    4.331121] i915 0000:00:02.0: IOV0: Initialization failed (-EIO) GT wedged
[    4.331126] i915 0000:00:02.0: [drm:i915_sriov_pf_confirm.cold [i915]] PF aborted (-EIO) i915_driver_probe+0x1b7f/0x1c50 [i915]
[    4.331300] i915 0000:00:02.0: No VFs could be associated with this PF!
[    4.331682] [drm] Initialized i915 1.6.0 20201103 for 0000:00:02.0 on minor 0
[    4.511122] sof-audio-pci-intel-tgl 0000:00:1f.3: bound 0000:00:02.0 (ops i915_audio_component_bind_ops [i915])
[    4.512869] fbcon: i915drmfb (fb0) is primary device
[    4.538902] i915 0000:00:02.0: [drm] fb0: i915drmfb frame buffer device`

**Comment by pwouik** [comment]
I am normally using "linux" arch packege(6.4.12.arch1-1 currently)

**Comment by right-0903** [comment]
> I am normally using "linux" arch packege(6.4.12.arch1-1 currently)
> maybe my linux-firmware is too recent

yours is too old, i915 out of tree is more recent than mainline, tgl latest guc in the git isn't 70.7

**Comment by right-0903** [comment]
> > I am normally using "linux" arch packege(6.4.12.arch1-1 currently)
> > maybe my linux-firmware is too recent
> 
> yours is too old, i915 out of tree is more recent than mainline, tgl latest guc in the git isn't 70.7, the firmware the module loaded is compatible downwardly, I don't want to explain here, you can read i915 source code.

**Comment by right-0903** [comment]
> > > I am normally using "linux" arch packege(6.4.12.arch1-1 currently)
> > > maybe my linux-firmware is too recent

 Yours is too old, the driver out of tree is more recent than mainline. And the firmware that the module loaded is compatible downwardly, I don't want to explain here, you can read i915 source code. Apparently, your module can't load firmware,  download firmware for tgl in the link I mentioned or log mentioned, and place them to `/lib/firmware`. just read log carefully, dude. `ERROR* GT0: GuC firmware i915/tgl_guc_70.7.0.bin: fetch failed -ENOENT`

**Comment by right-0903** [comment]
> I am normally using "linux" arch packege(6.4.12.arch1-1 currently)

You can't use it with Linux 6+ until it is merged to mainline, around Q4 2023. The dkms module for 6+ offered by this repository is incomplete, not all devices work, and can't wake from suspend.

**Comment by pwouik** [comment]
I managed to add all missing firmwares, I could echo in num_vfs
Thanks a lot for the help
Now I have error 43
windows 22H2
driver 4672

**Comment by right-0903** [comment]
> Now I have error 43
> windows 22H2
> with driver 4672 and 4577

code43 is really weird. my suggestion is try to add some quirks like vendor Id, use MS stock image with the latest update installed, remove virtual display like qxl, connect to it with RDP.

**Comment by pwouik** [comment]
turn out I just forgot vendor_id, mb

**Comment by right-0903** [comment]
> turn out I just forgot vendor_id, mb

ok, then if you make it work?

**Comment by pwouik** [comment]
I managed to use looking glass with iddsampledriver: https://github.com/ge9/IddSampleDriver
Everything works fine except minecraft bedrock
![image](https://github.com/strongtz/i915-sriov-dkms/assets/62726872/4f428699-219c-4282-af24-228463581c44)
also I had another screen in addition to the one I specified in the driver option.txt and I had to tell windows to only display on one of them to fix mouse position and have graphic acceleration

**Comment by cyril-cheng** [comment]
Can pve8 be installed?

**Comment by right-0903** [comment]
> I managed to use looking glass with iddsampledriver: https://github.com/ge9/IddSampleDriver
> Everything works fine except minecraft bedrock

glad to know, I only use it with RDP.

**Comment by right-0903** [comment]
> Can pve8 be installed?

A meaningfulless question, just test it out. I have mentioned I just tested with Manjaro.

**Comment by pagkly** [comment]
@right-0903 Finally I got it working. In my case, code 43 showed up due to me adding this to qemu .xml:
`    <feature policy="disable" name="hypervisor"/>`
so it wont work if someone tried to hide virtualization. 
After trying youtube 4k utilization remains 0% (but intel_gpu_top shows utilization) and Edge is not hardware accelerated.
![image](https://github.com/strongtz/i915-sriov-dkms/assets/28628331/99a03c0b-8adb-4d05-9d0b-9917a9f83a6e)
![image](https://github.com/strongtz/i915-sriov-dkms/assets/28628331/6c3218f3-08b7-422f-9d3e-ebb28b78f6b3)

![image](https://github.com/strongtz/i915-sriov-dkms/assets/28628331/35ff1b60-42e9-47e2-9b04-2adda83fc882)

**Comment by right-0903** [comment]
> @right-0903 Finally got it working after adding firmware. In my case, code 43 showed up due to me adding this line to qemu xml: ` <feature policy="disable" name="hypervisor"/>` so it wont work if someone tried to hide virtualization.

glad to know, I met this in one of my VMs, and I never realize it.

> After trying youtube 4k utilization remains 0% (but intel_gpu_top shows utilization) and Edge is not hardware accelerated.

if you are running a graphical target, then `intel_gpu_top` will be utilized, you should run it under headless, btw, how do you access the framebuffer of Vm? you should not use vm console. Use rdp instead, or you can use

> I managed to use looking glass with iddsampledriver: https://github.com/ge9/IddSampleDriver

**Comment by right-0903** [comment]
> I managed to use looking glass with iddsampledriver: https://github.com/ge9/IddSampleDriver
> Everything works fine except minecraft bedrock
> ![image](https://github.com/strongtz/i915-sriov-dkms/assets/62726872/4f428699-219c-4282-af24-228463581c44)
> also I had another screen in addition to the one I specified in the driver option.txt and I had to tell windows to only display on one of them to fix mouse position and have graphic acceleration

any idea with [intel zero copy virtual display driver](https://github.com/intel/Display-Virtualization-for-Windows-OS)?  

refer [this](https://open-iov.org/index.php/GPU_Driver_Internals#DVServer_KMD) too.

**Comment by sieskei** [comment]
Congratulations, @right-0903!
Am I correct in understanding that it only works on LTS kernel? In other words there is no possibility to compile it on 6.*.* kernels?

**Comment by right-0903** [comment]
> Am I correct in understanding that it only works on LTS kernel? In other words there is no possibility to compile it on 6.*.* kernels?

work with the version below Linux 6, you can see the patches for 4.x and 5.x in the backport repo. Linux 6 use maple tree for vma while Linux 5 use rb tree, there are many things to change. I do some changes, and compiled it for 6, but the module doesn't load.

**Comment by pwouik** [comment]
I tried intel 0 copy drivers: https://www.intel.com/content/www/us/en/download/737311/iotg-display-virtualization-drivers.html
but I dont know how to add a display
I tried to modify qemu commandline and video
[win10.xml.txt](https://github.com/strongtz/i915-sriov-dkms/files/12642471/win10.xml.txt)

**Comment by right-0903** [comment]
@pwouik 

> Associate QEMU with External Physical Display: For this we need to use customized QEMU version with "connectors" parameter enabled

It is sad to say that I have no extra monitor in my side, so I can't test it, I went to another branch of campus with only my laptop one month ago.

**Comment by pwouik** [comment]
I am talking about logical display, I have a laptop too

**Comment by right-0903** [comment]
> I am talking about logical display, I have a laptop too

0copy will give you a display adapter as mentioned, but you can't read it by ishmem due to zero copy, I don't know how can I access it if don't connect a monitor.

**Comment by right-0903** [comment]
@pwouik I have the ZC(zerocopy) driver 1314 built today, there is no huge difference for frames  between iddm and ZC when I do a furmark test, but that is very small even if it copys buffer under current resolution and frames update, maybe I should look into the overhead of cpu someday, I remember there is a overhead can't be ignored when do CPU benchmarking.

1. ZC with qemu commandline 
```bash
sudo qemu-system-x86_64 -m 8192 -enable-kvm -cpu host -smp cores=4,threads=2,sockets=1 \ 
-drive file=/mnt/win-vm/windows-vm/Win10-22h2.qcow2,format=qcow2,cache=none \
-device vfio-pci,host=0000:00:02.1 \
-drive file=/usr/share/edk2-ovmf/x64/OVMF_CODE.fd,if=pflash,format=raw,unit=0,readonly=on \
-drive file=~/.config/libvirt/qemu/nvram/win10-22h2_VARS.fd,if=pflash,format=raw,unit=1 \
-device virtio-vga,max_outputs=1,blob=true -display gtk,gl=on,full-screen=off \
-object memory-backend-memfd,id=mem1,size=8192M -machine memory-backend=mem1
```
display region keep flickering betwen the normal display and the blank background with text display is not active, and resolution can't be set greater than 1280x768.
![Screenshot_20231004_142244](https://github.com/strongtz/i915-sriov-dkms/assets/111179762/6c736dd2-7170-4cfc-86d7-5acd2ace9031)

2. ZC with lg and kvmfr module
works normally, resolution adjustment too
![Screenshot_20231004_142921](https://github.com/strongtz/i915-sriov-dkms/assets/111179762/4bf493e9-147a-4e3c-92a4-e5d5f55e0faa)

3. iddm with lg and kvmfr module
works normally, resolution adjustment too
![Screenshot_20231004_143604](https://github.com/strongtz/i915-sriov-dkms/assets/111179762/8205c5ed-92ba-4a84-82b6-9e06ffed9fa2)

4. virt-manager(spice) with ZC(just set display as virtio, no more  settings)
works normally, resolution adjustment too, ZC use iddm too when no monitor associated, but iddm have a conflict with the spice server, I cannot move touchpad when using `virt-manager`, but I can use keyboard.
![Screenshot_20231004_144206](https://github.com/strongtz/i915-sriov-dkms/assets/111179762/8255e577-3980-4c12-a9b0-2b78d598c760)

**Comment by right-0903** [comment]
some datasets for comparing cpu overhead

environment:

```
OS: Windows 10 22H2
vCPU: i7-11800h 4C8T
Ram: 8GB
vGPU: TigerLake-H GT1 (UHD Graphics)
```

power configuration sets: 

1. CPU governor: powersave PL1 45W PL2 55W(1s) PL4 68W 
2. CPU governor: performance PL1 90W PL2 95W(1s) PL4 200W

the result dominated by PL1(long time power without thermal throttling), PL4 maybe have a impact on current throttling

1. `virt-manager` with QXL can be seen as bare metal performance

![Screenshot_20231004_184133](https://github.com/strongtz/i915-sriov-dkms/assets/111179762/317e900c-c877-45fd-908b-2ef0f46f9a1a)
![Screenshot_20231004_183858](https://github.com/strongtz/i915-sriov-dkms/assets/111179762/9b4d8ec8-3d9c-4a57-8c46-8b7dd87d0b35)

2. qemu commandline (lack of optimization)
```bash
sudo qemu-system-x86_64 -m 8192 -enable-kvm -cpu host -smp cores=4,threads=2,sockets=1 \
-drive file=/mnt/win-vm/windows-vm/Win10-22h2.qcow2,format=qcow2,cache=none \
-device vfio-pci,host=0000:00:02.1 \
-drive file=/usr/share/edk2-ovmf/x64/OVMF_CODE.fd,if=pflash,format=raw,unit=0,readonly=on \
-drive file=/home/pengyu/.config/libvirt/qemu/nvram/win10-22h2_VARS.fd,if=pflash,format=raw,unit=1 \
-device virtio-vga,max_outputs=2,blob=true -display gtk,gl=on,full-screen=off \
-object memory-backend-memfd,id=mem1,size=8192M -machine memory-backend=mem1
```
![Screenshot_20231004_184630](https://github.com/strongtz/i915-sriov-dkms/assets/111179762/9a644afd-d1c8-46aa-b1c7-5c5718e307fb)
![Screenshot_20231004_184936](https://github.com/strongtz/i915-sriov-dkms/assets/111179762/6e24ede8-bbb2-48d5-be28-4399e868b11f)

3. ZC with lg and kvmfr module
![Screenshot_20231004_185412](https://github.com/strongtz/i915-sriov-dkms/assets/111179762/a4d4111d-9a67-4636-b347-5848d4311cb8)
![Screenshot_20231004_185329](https://github.com/strongtz/i915-sriov-dkms/assets/111179762/72c91718-82fc-4733-affa-84fb1480e58e)

4. ZC
![Screenshot_20231004_191926](https://github.com/strongtz/i915-sriov-dkms/assets/111179762/cfd22b9b-3b8f-4934-a265-4becb7f0faf3)
![Screenshot_20231004_191847](https://github.com/strongtz/i915-sriov-dkms/assets/111179762/8e32255c-507b-4870-b047-f1a9ce20caf9)

**Comment by rbragadeandrade** [comment]
@right-0903 thanks very much for the tip, when using backport with 5.15 kernel sr-iov now works for my i7-1165g7. The only issue I encoutered is that if I enable VF when I try to shutdown the PC or reboot it hangs, this happens even if I disable the VF before rebooting or shutdown. Does this happen to you also?

**Comment by right-0903** [comment]
> @right-0903 thanks very much for the tip, when using backport with 5.15 kernel sr-iov now works for my i7-1165g7. The only issue I encountered is that if I enable VF when I try to shutdown the PC or reboot it hangs, this happens even if I disable the VF before rebooting or shutdown. Does this happen to you also?

never, maybe you should look into `journalctl -rb -1`, BTW your kernel info?

**Comment by right-0903** [comment]
It took me some time to modify it, finally, I make it works with linux 6, I think we don't need many changs to make it work with linux6.5.

![Screenshot_20231008_182920](https://github.com/strongtz/i915-sriov-dkms/assets/111179762/48b24f89-d5df-472a-a04a-6f8455b016b4)

**Comment by rba-andrade** [comment]
Can you share your patches?

**Comment by right-0903** [comment]
> At the moment, I'm using Fedora 38 6.5.6. Will your fork work with it?

I use 6.5.5-zen daily, I can't build the module with it. but we can patch it according to mainline kernel source, since the obstacle from backport have been removed now. Recently, I have no much time to do it. Maybe you can dive into it.

**Comment by resiliencer** [comment]
It can be build for 6.5.7-zen or 6.5.9-zen in arch, but a memory leak occurs and host will run out of memory in hour or two. Error in dmesg: `[drm] *ERROR* tlb invalidation response timed out for seqno 23`.
I didn't investigate this issue too much, currently I'm using 6.1.60-lts and it works perfectly. Thanks for the ideas from this thread. They helped me a lot. Especially @right-0903. Finally managed to make it all work on my raptor lake with only single iGPU via SR-IOV and Looking Glass for win10/win11.

**Comment by purplepalmdash** [comment]
gen 11 works perfect using this "intel-i915-dkms bakckport ", while in gen 12(i3-12100, uhd730) I always get code 43 in vm.

**Comment by right-0903** [comment]
> gen 11 works perfect using this "intel-i915-dkms bakckport ", while in gen 12(i3-12100, uhd730) I always get code 43 in vm. 

many reasons cause code 43, like hiding virtualization, improper driver, etc. You should submit more information. BTW, i915 backport have support linux 6.0lts officially, you can also report this to Intel.

**Comment by sieskei** [comment]
Hello, all!
Just now, I find time for it and I managed to compile it for LTS 6.2.
I'm trying it with Win11, and I think it's working, but I'm having trouble launching the Looking Glass host.
I created a virtual monitor with IDDM.
What are the settings in your Looking Glass INI file?

**Comment by right-0903** [comment]
> Hello, all!
> Just now, I find time for it and I managed to compile it for LTS 6.2.
> I'm trying it with Win11, and I think it's working, but I'm having trouble launching the Looking Glass host.
> I created a virtual monitor with IDDM.
> What are the settings in your Looking Glass INI file?

You can create a qxl display to check, I don't set specific settings. BTW, 6.2 is not LTS, even though 6.2 is supported officially now.

**Comment by sieskei** [comment]
```
[I]    335250769              time.c:85   | windowsSetTimerResolution      | System timer resolution: 500.0 μs
[I]    335251612               app.c:809  | app_main                       | Looking Glass Host (B6-225-22d949c4)
[I]    335252430           cpuinfo.c:38   | cpuInfo_log                    | CPU Model: 11th Gen Intel(R) Core(TM) i9-11900KB @ 3.30GHz
[I]    335253241           cpuinfo.c:39   | cpuInfo_log                    | CPU: 1 sockets, 6 cores, 12 threads
[I]    335254847           ivshmem.c:132  | ivshmemInit                    | IVSHMEM 0* on bus 0x6, device 0x2, function 0x0
[I]    335266380               app.c:826  | app_main                       | IVSHMEM Size     : 64 MiB
[I]    335266610               app.c:827  | app_main                       | IVSHMEM Address  : 0x2753BEA0000
[I]    335266846               app.c:828  | app_main                       | Max Pointer Size : 1024 KiB
[I]    335267074               app.c:829  | app_main                       | KVMFR Version    : 20
[I]    335267292               app.c:848  | app_main                       | Trying           : DXGI
[I]    335269957              dxgi.c:448  | dxgi_init                      | Not using unsupported adapter: Microsoft Basic Render Driver
[I]    335270368              dxgi.c:448  | dxgi_init                      | Not using unsupported adapter: Microsoft Basic Render Driver
[E]    335270681              dxgi.c:503  | dxgi_init                      | Failed to locate a valid output device
[I]    335270961               app.c:848  | app_main                       | Trying           : NVFBC
   335271567 [E]          wrapper.cpp:61   | NvFBCInit                      | Failed to load NvFBC64.dll: 0x0000007e (The specified module could not be found.)
[E]    335271947               app.c:868  | app_main                       | Failed to find a supported capture interface
[I]    335273971               app.c:1029 | app_main                       | Host application exited
```

![Screenshot 2023-11-21 145753](https://github.com/strongtz/i915-sriov-dkms/assets/58069642/1e57c793-867b-4b6a-855c-ca4758874aca)

**Comment by sieskei** [comment]
It's working now. The issue was that the screen was set to clone mode.

**Comment by right-0903** [comment]
> It's working now. The issue was that the screen was set to clone mode.

You mean windows setting? I never meet that.

**Comment by purplepalmdash** [comment]
> > gen 11 works perfect using this "intel-i915-dkms bakckport ", while in gen 12(i3-12100, uhd730) I always get code 43 in vm.
> 
> many reasons cause code 43, like hiding virtualization, improper driver, etc. You should submit more information. BTW, i915 backport have support linux 6.0lts officially, you can also report this to Intel.

Tried the same configuration(use the same ssd for booting 2 machines) on different platform(gen11 vs gen 13), following is the result. 

| OS/Hardware        | Gen 13(i5-13400)/(uhd 730)           | Gen 11(i5-1135G7)/(Iris Xe Graphics) | dkms Repository  |
| ------------- |:-------------:|:-------------:| -----:|
| RockLinux 5.14.0-284(official)/Qemu7.1      | Code 43, gpu not working in vm | OK | https://github.com/intel-gpu/intel-gpu-i915-backports |
| RockLinux 6.1.63(self-build) /Qemu7.1     | OK      | Code 43, can see vf in vm but not working |  https://github.com/strongtz/i915-sriov-dkms |


Just wondering what caused this difference(cpu gen? iGPU type?), does anybody runs well using backport driver on gen12/13?

**Comment by right-0903** [comment]
@purplepalmdash hope you have read all the thread. Then dive into those logs.

**Comment by plantroon** [comment]
1235U here, error 43. Gentoo with kernel 6.1 LTS. But I could get it working with https://github.com/strongtz/i915-sriov-dkms - however this one makes the laptop freeze when entering sleep mode.

**Comment by right-0903** [comment]
> 1235U here, error 43. Gentoo with kernel 6.1 LTS. But I could get it working with https://github.com/strongtz/i915-sriov-dkms - however this one makes the laptop freeze when entering sleep mode.

Let's wait for xe driver, which supports sriov natively, coming in linux 6.8

**Comment by pelhadda** [comment]
> I managed to use looking glass with iddsampledriver: https://github.com/ge9/IddSampleDriver Everything works fine except minecraft bedrock ![image](https://private-user-images.githubusercontent.com/62726872/266257767-4f428699-219c-4282-af24-228463581c44.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMyODEyNjUsIm5iZiI6MTcwMzI4MDk2NSwicGF0aCI6Ii82MjcyNjg3Mi8yNjYyNTc3NjctNGY0Mjg2OTktMjE5Yy00MjgyLWFmMjQtMjI4NDYzNTgxYzQ0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjIyVDIxMzYwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIxZTI3ZTU4OGI1YWM1ZjMwOGI0MGRhNmQ4ZTJjODIwZDU1OGY2NGFjMTQxNjIwNzE1MGFkYTM4MzE1ZmI1NDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1kIC7tT7Jmr2ob6OGPIApzzTfEEGNYiz-LrGp4DkriY) also I had another screen in addition to the one I specified in the driver option.txt and I had to tell windows to only display on one of them to fix mouse position and have graphic acceleration

I had the same issue with Minecraft. In Windows 11 Settings > Display > Graphics, and set Minecraft to "Specific GPU".

![image](https://github.com/strongtz/i915-sriov-dkms/assets/66429828/acde649a-65a4-4bd8-98ac-7f43cc5bf1ff)

**Comment by rba-andrade** [comment]
Has anyone else tried this with xe kernel driver from 6.8 mainline kernel? I get "xe 0000:00:02.0: driver does not support SR-IOV configuration via sysfs" when trying to enable 1 VFs.

**Comment by plantroon** [comment]
> Has anyone else tried this with xe kernel driver from 6.8 mainline kernel? I get "xe 0000:00:02.0: driver does not support SR-IOV configuration via sysfs" when trying to enable 1 VFs.

You're not using the Xe driver but i915 (as that's still the default) you must use xe.force_probe to use Xe driver, like: `i915.force_probe=!56a2 xe.force_probe=56a2` where 56a2 is your GPU's model ID from `lspci -nn`. 

I did try, good luck getting even a desktop running. Long way to go.

**Comment by rba-andrade** [comment]
I did use the force probe, I can get desktop running, but no sucess with SR-IOV. You need to compile mesa with meson build option "-Dintel-xe-kmd=enabled" so you can get to desktop.

**Comment by plantroon** [comment]
I built mesa-git with that build option and it only showed me rainbows or psychedelic imagery in all sddm, gdm, hyprland and gnome session. Is a specific version of mesa needed?

**Comment by plantroon** [comment]
Tried mesa 24.0.1 as well. In both cases I get errors like in the photoscreenshot.
![20240224_115626](https://github.com/strongtz/i915-sriov-dkms/assets/9124099/2e7f58e3-7dba-4589-90cf-4f2f1d9a4776)
😔

**Comment by rba-andrade** [comment]
Didn't work for me with mesa-git, but mesa 24.0.1 works. I'm using an i7-1165G7 with Iris Xe graphics.

**Comment by plantroon** [comment]
Yes CONFIG_DRM_XE_DISPLAY=y was configured. How did mesa-git behave on your side?

**Comment by rba-andrade** [comment]
Has anyone successfully built the new version of the i915-backport which enables support for kernel 6.5 against the 6.6 lts?

**Comment by right-0903** [comment]
> Has anyone successfully built the new version of the i915-backport which enables support for kernel 6.5 against the 6.6 lts?

yes, I gave it a try, after did some changes, it can be built, then boot into desktop, throw dereference null ptr error, and drm issue, screen freezing. nowadays, it's a mess, they backported drm module, added dmabuf module, give several build option i915, i915_only, drm, dmabuf, nodrm and so on, but I don't find a doc about how to really do it. I set build all modules, but it wouldn't.

**Comment by right-0903** [comment]
at your risks, a rough and unstable patch here.
![Screenshot_20240309_223805](https://github.com/strongtz/i915-sriov-dkms/assets/111179762/85bc7fea-db58-4e35-97ca-9d0123dc42d7)
[i915-out-tree-for-Linux6.6-progressing.txt](https://github.com/strongtz/i915-sriov-dkms/files/14548830/i915-out-tree-for-Linux6.6-progressing.txt)

**Comment by rba-andrade** [comment]
Which defconfig did you use? I'm getting stuck at loading modules during init.

**Comment by right-0903** [comment]
> Which defconfig did you use? I'm getting stuck at loading modules during init.

i915, but I tried all deconfigs, then I got the same .config

**Comment by rba-andrade** [comment]
Did anyone already tried the new backport version with kernel 6.6 support (I915_24WW23.5_881.19_24.2.17_240301.20)? I'm trying it with kernel 6.6.32 and is not working well for me. It only boots with "ibt=off" kernel parameter because of MISSING ENDBR instructions. Is there any changes to compile parameters that makes it work?

**Comment by zyxhere** [comment]
Hi chiming in, am total newbie just wanna ask some questions (sorry for clutter) Can sr-iov work out of the box now for example on debian unstable which has kernel 6.8 (has the new xe driver which supposedly has it sriov built in?) and mesa 24.1.1? if no then what components are still missing? Also do you need anything other than virtualization enabled in bios? I have a basic locked down bios on a 12th gen laptop (i5-1235U)

**Comment by johntdavis84** [comment]
Hello,

These are great questions. :) There’s another GitHub where the work is being done to get this into the mainline Linux kernel, but I’m not sure which repo that is.

SR-IOV features with the i915 requires this DKMS driver (or the one from Intel’s repo—I’m not sure which one is further along in development right now) be manually installed and maintained. 

SR-IOV will not work out of the box with the Linux kernel until the replacement for the i915 driver (the upcoming “Xe” driver) replaces the i915 driver as the default, which is likely not to happen until, I believe, kernel 6.11, which might be released in October, and will then need to filter down to distro releases.

It’s a little confusing because the Xe driver is a huge change, and they’ve been implementing a lot of the underpinning infrastructure since at least 6.8. Xe will appear as a non-default replacement for i915 before it becomes the default, most likely.

Keep in mind that when SR-IOV hits mainline, you won’t need this DKMS or have to do any of this, but that won’t be until at least October.

If your BIOS lets you enable virtualization enough for, e.g., Proxmox to work, you should be fine.
Some BIOSes have a separate option to enable SR-IOV or VTd, but if your BIOS doesn’t have those, assume they were automatically turned on when you enabled virtualization, unless you’re having issues.

One thing you absolutely need to do if you can is disable secure boot. It makes this all more difficult and unpredictable.

There are some tutorials that show you how to verify you have your BIOS settings correct. The first two links assume you're running proxmox, but the under the hood setup and verification of the BIOS settings are the same for any Debian/Ubuntu.
https://pve.proxmox.com/wiki/PCI(e)_Passthrough￼
https://pve.proxmox.com/wiki/PCI_Passthrough￼
https://www.derekseaman.com/2023/11/proxmox-ve-8-1-windows-11-vgpu-vt-d-passthrough-with-intel-alder-lake.html￼
Proxmox VE 8.2: Windows 11 vGPU (VT-d) Passthrough with Intel Alder Lake -
derekseaman.com


					- J.

**Comment by r0nn** [comment]
Thanks, my intel nuc11pah i5 works on Proxmox 8.2 kernel 6.8.8-4-pve (not working with kernel 6.8.12-1 due to some kernel changes), using most recent commit of backport driver (I915_24WW33.3_950.13_24.4.12_240603.18), but need some tweak in dkms.conf:

```
BUILT_MODULE_NAME[0]="compat"
BUILT_MODULE_LOCATION[0]="compat"
DEST_MODULE_LOCATION[0]="/updates"
```
should change to 
```
BUILT_MODULE_NAME[0]="i915-compat"
BUILT_MODULE_LOCATION[0]="compat"
DEST_MODULE_LOCATION[0]="/updates"
```

that's all.

**Comment by DreamChaser-luzeyu** [comment]
Thanks for your info! I can confirm this works for my i5 1135g7 processor! I believe this is the correct approach for tiger lake.
I am running Ubuntu 24.04 with kernel 6.8 which is officially supported by the backport driver. They also says kernel 6.11 will work.

In my test kernel 6.12 on debian 13 does not work, the kernel panic on boot.

<img width="2095" height="1356" alt="Image" src="https://github.com/user-attachments/assets/82c526b0-e1c2-4a67-b0ae-a098196d9097" />
